catalog/resource_grant.go

```diff
@@ -0,0 +1,187 @@
+			Principal: principal,
+			Remove:    util.SetToSlice(remove),
+		})
+	}
+	// so that we can deterministic tests
+	sort.Slice(diff.Changes, func(i, j int) bool {
+		return diff.Changes[i].Principal < diff.Changes[j].Principal
+	})
+	return diff
+}
+
+// replacePermissionsForPrincipal merges removal diff of existing permissions on the platform
+func replacePermissionsForPrincipal(a permissions.UnityCatalogPermissionsAPI, securable string, name string, principal string, list permissions.UnityCatalogPermissionsList) error {
+	existing, err := a.GetPermissions(securable, name)
+	if err != nil {
+		return err
+	}
+	return a.UpdatePermissions(securable, name, diffPermissionsForPrincipal(principal, list, existing))
+}
+
+func filterPermissionsForPrincipal(in permissions.UnityCatalogPermissionsList, principal string) (out permissions.UnityCatalogPermissionsList) {
+
+	for _, v := range in.Assignments {
+		if v.Principal == principal {
+			out.Assignments = append(out.Assignments, v)
+		}
+	}
+	return
+}
+
+func ResourceGrant() *schema.Resource {
+	s := map[string]*schema.Schema{
+		"principal": {
+			Type:     schema.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
+		"privileges": {
+			Type:     schema.TypeSet,
+			Required: true,
+			Elem:     &schema.Schema{Type: schema.TypeString},
+			Set:      schema.HashString,
+		},
+	}
+	// Handle all of the securable resource types
+	allFields := []string{}
+	for field := range permissions.Mappings {
+		allFields = append(allFields, field)
+	}
+
+	for _, field := range allFields {
+		s[field] = &schema.Schema{
+			Type:          schema.TypeString,
+			Optional:      true,
+			ForceNew:      true,
+			ConflictsWith: util.SliceWithoutString(allFields, field),
+		}
+	}
+
+	return common.Resource{
+		Schema: s,
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
+			if d.Id() == "" {
+				// unfortunately we cannot do validation before dependent resources exist with tfsdkv2
+				return nil
+			}
+			var grants permissions.UnityCatalogPermissionsList
+			common.DiffToStructPointer(d, s, &grants)
+			return permissions.Mappings.Validate(d, grants)
+		},
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			principal := d.Get("principal").(string)
+			privileges := util.ToStringSlice(d.Get("privileges").(*schema.Set).List())
+			var grants = permissions.UnityCatalogPermissionsList{
+				Assignments: []permissions.UnityCatalogPrivilegeAssignment{
+					{
+						Principal:  principal,
+						Privileges: privileges,
+					},
+				},
+			}
+			securable, name := permissions.Mappings.KeyValue(d)
+			unityCatalogPermissionsAPI := permissions.NewUnityCatalogPermissionsAPI(ctx, c)
+			err := replacePermissionsForPrincipal(unityCatalogPermissionsAPI, securable, name, principal, grants)
+			if err != nil {
+				return err
+			}
+			d.SetId(permissions.Mappings.Id(d))
```

Comment:
this id won't be unique - we should use the principal in the id

Reply:
Done